geps/gep-713.md

```diff
@@ -277,18 +444,46 @@ spec:
   default:
     maxRetries: 5
   targetRef:
+    type: direct
     group: networking.acme.io
     kind: ExternalService
     name: foo.com
 ```
 
+### Merge semantics
+
+It's possible (even likely) that configuration in a Policy may need to be merged
+into an existing object somehow, particularly for Hierarchical policies.
+
+In general, Policy objects should merge values at a scalar level, not at a
+whole-struct level.
+
+For example, in the `GKEServicePolicy` example above, the `cdn` struct contains
+a `cachePolicy` struct that contains fields. If an implementation was merging
+this configuration into an existing object that contained the same fields, it
+should merge the fields at a scalar level, with the `includeHost`,
+`includeProtocol`, and `includeQueryString` values being defaulted if they were
+not specified in the object being controlled. Similarly, for `overrides`, the
+values of the innermost scalar fields should overwrite the scalar fields in the
+affected object.
+
+Implementations should not copy the higher-level structs directly into the
+affected object.
+
+In the case that the field in the Policy affects a struct that is a member of a list,
```

Comment:
lets take an example of a Hierarchical policy
* Only specify defaults
Applied at Gateway
```
RetryPolicy
default:
  retryOn:
  - "501"
  - "502"
  - "503"
```
Applied at xRoute
```
RetryPolicy
default:
  retryOn:
  - "504"
```
is this the final state (appending/merging `defaults`) ?
```
retryOn:
  - "501"
  - "502"
  - "503"
  - "504"
 ``` 
 
* Top level specifies a default and child level specifies a override
Applied at Gateway
```
RetryPolicy
default:
  retryOn:
  - "501"
  - "502"
  - "503"
```
Applied at xRoute
```
RetryPolicy
override:
  retryOn:
  - "504"
```
is this the final state (bottom level `override` wins) ?
```
retryOn:
  - "504"
 ``` 
 * Top level specifies a override and child level specifies a default
Applied at Gateway
```
RetryPolicy
override:
  retryOn:
  - "501"
  - "502"
  - "503"
```
Applied at xRoute
```
RetryPolicy
default:
  retryOn:
  - "504"
```
is this the final state (top level `override` wins) ?
```
retryOn:
  - "501"
  - "502"
  - "503"
 ```

Reply:
Just my 2c, it probably makes sense to be implementation specific how merging is done. For this case, I would recommend implementations replace.

If we do prescribe replace vs merge it should be replace, IMO, since that covers more use cases (you can always add more with replace, but cannot take away with merge).

Reply:
yeah agreed `replace` for all cases isa good solution, would prefer if this was a recommendation from this project

Reply:
Okay, there's some confusion here about the _intent_ of `defaults` and `overrides`, and it's also intersecting with a particular type of element that makes merging very difficult.

For the latter, merging lists is always going to be hard, that's why Kubernetes (and kubebuilder) have the listMergeType directives, so that the apiserver and other api machinery get some clues about how to merge things. A list of strings is probably close to the worst-case for merge complexity, because there are no clues about priority _in the string itself_, unlike if there was a structured object there.

That said, I can see that in the example here, it does make sense. But we need to understand that we should _try_ to avoid lists-of-strings in the same way we should avoid `map[string]string` - the lack of structure makes doing things with the field harder in the long term. (That isn't to say we should never do it, just that it has a cost we should consider).

For this specific setting, you could achieve the same aim by using a single string that contained comma-separated values. To that end, when a Policy changes the `retryOn` field, it will _replace_ the field, not be merged into it. Effectively, the _value_ of the list is the _set of strings_ included in the list, and the Policy will replace the value, the same as it would on a `key: value` field inside a struct. The same applies for `value: map[string]string` fields.

In this case, the other rule that's important is the differences between `defaults` and `overrides`.

The overall rule is that only one instance of a Policy should be in effect on the final object at any one time. If two Policy objects of the same type end up affecting the same leaf node, then only one can win.

`defaults` are a "more specific wins" construct, so if multiple Policy objects are attached, the one that's _closer to the thing being modified_ will win.

`overrides` are a "less specific wins" construct, so if multiple Policy objects are attached, the ones that's _higher in the hierarchy_ will win.

In the event a default and an override both apply, no matter what level they've come from, then the override will win.

This means that in the first example, the defaults from the two Policy objects should _not_ be merged. The defaults should be merged _into_ the thing they're modifying, but only one instance of the same Policy will be the winner.
That means that the final result should look like

```
RetryPolicy
default:
  retryOn:
  - "504"
```

This is the same as if the actual Route had specified the setting.

For the second example, top level specifies a default and a lower an override, the override wins, because overrides beat defaults.

For the third, top level specifies an override, and a lower a default, the override wins, because overrides beat defaults.

Some other examples:
Top level specifies an override, and a lower a different override, the top-level override wins, because overrides are "less specific beats more specific".

This comment was me working through my thoughts, I'll take it and fold it into the GEP.

Reply:
I'll leave this thread here for a bit so everyone can see it, but I've folded this language into both the GEP and the doc page now, so I consider this thread resolved.